### ISSUE TITLE
Small chance to compile wrong page with ruby 2.7

### ISSUE THREAD

**Issue by Fjan** [nanoc/nanoc#1501]
Under Ruby 2.7 there is a very small chance that nanoc compile will use the wrong content for a page when a large number of pages change at the same time

### Steps to reproduce

1. use ruby-2.7
2. Make a small change that affects all pages, like adding a single character to the layout
3. `nanoc compile`
4. Clear the cache to force a recompile `rm -rf tmp`
5. `nanoc compile --diff`

### Expected behavior

The second compile should show no changes.

### Actual behavior

About 2 to 4 of every 1000 pages changes, to the correct version. Subsequent `nanoc compile` commands show no changes.

### Details

It seems to work OK on ruby 2.6 and it also works fine on smaller repos and smaller changes. Only during a compile that affects 1000 pages does the compilation get confused.

**Comment by ddfreyne** [comment]
This bug is disturbing, and even more disturbing is that I am unable to reproduce it :(

@Fjan Can you check something for me? Drop the code snippet below into your site’s `lib/` directory (location does not matter), and then try to reproduce the issue again. Can you reproduce it with this code snippet in place?

```ruby
module Nanoc
  module Core
    module CompilationPhases
      class Write < Abstract
        include Nanoc::Core::ContractsSupport

        def initialize(compiled_content_store:, wrapped:)
          super(wrapped: wrapped)

          @compiled_content_store = compiled_content_store
        end

        def start
          super
        end

        def stop
          super
        end

        contract Nanoc::Core::ItemRep, C::KeywordArgs[is_outdated: C::Bool], C::Func[C::None => C::Any] => C::Any
        def run(rep, is_outdated:) # rubocop:disable Lint/UnusedMethodArgument
          yield

          # Caution: Notification must be posted before enqueueing the rep,
          # or we risk a race condition where the :rep_write_ended
          # notification happens before the :rep_write_enqueued one.
          Nanoc::Core::NotificationCenter.post(:rep_write_enqueued, rep)

          writer = Nanoc::Core::ItemRepWriter.new
          writer.write_all(rep, @compiled_content_store)
        end
      end
    end
  end
end
```

**Comment by Fjan** [comment]
Yes, I believe that fixes it. I did about ten compiles with the patch, each time changing a single character and then doing two compiles, and it appears to work flawlessly.

**Comment by ddfreyne** [comment]
#1507 should fix this!

**Comment by Fjan** [comment]
Disappointingly, I still managed to get an incorrect compile once when using nanoc 4.11.18 on ruby 2.7. It's hard to reproduce though, it takes a lot of recompiles to get one that goes wrong (on an 8-core iMac Pro).

**Comment by ddfreyne** [comment]
@Fjan Does it happen with the code snippet mentioned above, too?

**Comment by Fjan** [comment]
I wrote a little script to automate the problem of changing a character in the layout and compiling twice to help detect the issue. I first tried it without the patch. I let that run for half an hour and... wasn't able to reproduce it at all anymore. I would swear I've seen it happen once when I did it manually but also several manual attempts no longer revealed the issue. So I suggest closing this for now until I find a reliable way to reproduce it again.

**Comment by Fjan** [comment]
I've re-opened this issue. Over the last month it has happened to me twice that items got compiled to the wrong name. A recompile fixed it, but it's still rather serious if you don't happen to notice the change. So the issue is very rare, but there is still a race condition in nanoc somewhere that gets exposed by Ruby 2.7.

**Comment by Ravlen** [comment]
@ddfreyne @Fjan While I can't reproduce this consistently, we just recently updated to Ruby 2.7 and now we are starting to see random failures in our CI pipeline, which compiles ~1300 files each run. For example: https://gitlab.com/gitlab-org/gitlab/-/jobs/791090474#L2750. The pipeline job runs many times a day, so we have a lot more potential for running into this. Rerunning the job seems to clear it every time.

In the error I listed above, the file causing the error only has two lines, and most certainly does not have the 4 links reported in the error: https://gitlab.com/gitlab-org/gitlab/-/tree/master/doc/user/project/status_page. Nanoc seems to have compiled it with the content from a different file. I didn't dig into it to see which file it took the contents from, but I can try to figure it out if that helps.

**Comment by ddfreyne** [comment]
Oh, I had naïvely assumed that the issue would lie with the caching, but now I see that it also happens with a fresh compilation.

Now, if only I could reproduce it still. I have managed to get the error twice myself locally, but never again since.

@Ravlen If you apply the patch suggested in https://github.com/nanoc/nanoc/issues/1501#issuecomment-643439217, does the problem go away?

**Comment by Fjan** [comment]
@ddfreyne The chance seems higher with a fresh compilation, I don't think I've seen it happen when only a few pages change. But that could be just because the chance is dependent on the number of pages that change.

**Comment by ddfreyne** [comment]
Yep — I can reproduce it now, but not with the patch in the comment. Can anyone confirm? (I’m running a long-term test now.)

If that patch actually fixes the problem, then my task is to figure out _why_… which is still a mystery to me.

**Comment by Fjan** [comment]
Ok, I've added the patch to our repo (minus the `include Nanoc::Core::ContractsSupport` line since that resulted in warnings). But it's going to take a while to build confidence that it fixes it, because the issue is intermittent.

**Comment by ddfreyne** [comment]
I believe I have found the culprit: a non-threadsafe object used across threads. A fix is in #1510.

More proof that concurrency is not for the faint of heart.

**Comment by ddfreyne** [comment]
Will release a fix for this soon.

**Comment by ddfreyne** [comment]
This is now fixed in the just-released 4.11.19. (It might take a few minutes before that release becomes available for installing.)
